{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import anndata as ad\n",
    "from scipy.stats import spearmanr\n",
    "import scanpy.external as sce\n",
    "from itertools import product"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "path_cwd=os.getcwd()\n",
    "path_train_data=path_cwd + '/ML4G_Project_2_Data/train_data'\n",
    "path_test_data=path_cwd + '/ML4G_Project_2_Data/test_data'\n",
    "#train_data\n",
    "\n",
    "bulk_train=pd.read_csv(path_train_data +'/pancreas_bulk_train.csv',index_col=0)\n",
    "metadata_train=pd.read_csv(path_train_data+ '/pancreas_sc_metadata_train.csv')\n",
    "sc_train=pd.read_csv(path_train_data +'/pancreas_sc_train.csv', index_col=0)\n",
    "\n",
    "#test_data\n",
    "metadata_test=pd.read_csv(path_test_data+ '/pancreas_sc_metadata_test_wocelltype.csv')\n",
    "sc_test=pd.read_csv(path_test_data +'/pancreas_sc_test.csv', index_col=0)\n",
    "\n",
    "cells_1 = [column for column in sc_train.columns if 'patient1' in column ]\n",
    "cells_2 = [column for column in sc_train.columns if 'patient2' in column ]\n",
    "cells_3 = [column for column in sc_train.columns if 'patient3' in column ]\n",
    "cells_4 = [column for column in sc_train.columns if 'patient4' in column ]\n",
    "\n",
    "# we transpose the dataframe since we want a cells x genes dataframe for applying MAGIC\n",
    "sc1_train=sc_train[cells_1].T\n",
    "sc2_train=sc_train[cells_2].T\n",
    "sc3_train=sc_train[cells_3].T\n",
    "sc4_train=sc_train[cells_4].T\n",
    "\n",
    "# we aggregate the dataframes in a list\n",
    "patients=[sc1_train,sc2_train,sc3_train,sc4_train]\n",
    "\n",
    "# we extract genes that are not constant across cells\n",
    "genes1=list(sc1_train.loc[:, (sc1_train != sc1_train.iloc[0]).any()].columns)\n",
    "genes2=list(sc2_train.loc[:, (sc2_train != sc2_train.iloc[0]).any()].columns)\n",
    "genes3=list(sc3_train.loc[:, (sc3_train != sc3_train.iloc[0]).any()].columns)\n",
    "genes4=list(sc4_train.loc[:, (sc4_train != sc4_train.iloc[0]).any()].columns)\n",
    "\n",
    "genes=[genes1,genes2,genes3,genes4]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Pre processing: Normalize per total reads per cell and transform with square root\n",
    "\n",
    "# Helper function to pre process data\n",
    "def pp(data):\n",
    "    data=ad.AnnData(data) # needed to use following functions\n",
    "    #sc.pp.filter_cells(data,min_counts=1)\n",
    "    sc.pp.normalize_total(data)\n",
    "    sc.pp.sqrt(data)\n",
    "    data=pd.DataFrame(data.X, columns = data.var_names, index=data.obs_names)\n",
    "    return data\n",
    "\n",
    "# we preprocess the data for each patient\n",
    "patients_pp=[]\n",
    "for i,data in enumerate(patients):\n",
    "    pp_data=pp(data)\n",
    "    patients_pp += [pp_data]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Using the whole train dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "parameters: (60, 'auto', 175)\n",
      "Correlation After imputation: 0.9470659901458889\n",
      "--------\n",
      "parameters: (60, 'auto', 180)\n",
      "Correlation After imputation: 0.9470291760147787\n",
      "--------\n",
      "parameters: (60, 'auto', 185)\n",
      "Correlation After imputation: 0.9470344100812496\n",
      "--------\n",
      "parameters: (80, 'auto', 175)\n",
      "Correlation After imputation: 0.9473971744862868\n",
      "--------\n",
      "parameters: (80, 'auto', 180)\n",
      "Correlation After imputation: 0.9473676454934784\n",
      "--------\n",
      "parameters: (80, 'auto', 185)\n",
      "Correlation After imputation: 0.9473711161147803\n",
      "--------\n",
      "parameters: (90, 'auto', 175)\n",
      "Correlation After imputation: 0.9474921757883071\n",
      "--------\n",
      "parameters: (90, 'auto', 180)\n",
      "Correlation After imputation: 0.9474628878062937\n",
      "--------\n",
      "parameters: (90, 'auto', 185)\n",
      "Correlation After imputation: 0.9474705820892478\n",
      "--------\n",
      "parameters: (100, 'auto', 175)\n",
      "Correlation After imputation: 0.9475472372512046\n",
      "--------\n",
      "parameters: (100, 'auto', 180)\n",
      "Correlation After imputation: 0.947519650337976\n",
      "--------\n",
      "parameters: (100, 'auto', 185)\n",
      "Correlation After imputation: 0.9475264271048306\n",
      "--------\n",
      "Correlation before imputation: 0.9358073816280966\n",
      "BEST CORRELATION AFTER IMPUTATION: 0.9475472372512046\n",
      "best parameters: [100, 'auto', 175]\n"
     ]
    }
   ],
   "source": [
    "# concatenating the datasets\n",
    "data_complete=pd.concat(patients_pp,axis=0)\n",
    "genes=list(data_complete.loc[:, (data_complete != data_complete.iloc[0]).any()].columns)\n",
    "lens=[len(x) for x in patients_pp]\n",
    "indexes=[np.sum(lens[:i+1]) for i in range(4)]\n",
    "results=[]\n",
    "\n",
    "# hyperparameter search space for magic\n",
    "k_set ,t_set, n_pca_set= [60,80,90,100], ['auto'] ,[175,180,185]\n",
    "best_params=0\n",
    "best_score=0\n",
    "\n",
    "# imputation between single cell data and bulk data (baseline)\n",
    "baseline=np.mean([spearmanr(patients_pp[i].mean(axis=0),bulk_train['patient' + str(i+1)]).statistic for i in range(4)])\n",
    "\n",
    "# MAGIC imputation with hyperparameter search\n",
    "for k,t,n_pca in product(k_set,t_set,n_pca_set):\n",
    "    imputed_global=data_complete.copy()\n",
    "    imputed_local=sce.pp.magic(ad.AnnData(data_complete.loc[:,genes]),knn=k,t=t,n_pca=n_pca,n_jobs=8,copy=True,verbose=0,solver='approximate',random_state=7)\n",
    "    imputed_global.loc[:,genes]=imputed_local.X\n",
    "    result=0\n",
    "    for i,index in enumerate(indexes):\n",
    "        if i==0: sc_imputed= imputed_global.iloc[:index,:]\n",
    "        else: sc_imputed= imputed_global.iloc[indexes[i-1]:index,:]\n",
    "        bulkified=sc_imputed.mean(axis=0)\n",
    "        bulk=bulk_train['patient' + str(i+1)]\n",
    "        result+=spearmanr(bulkified,bulk).statistic/4\n",
    "    if result >= best_score:\n",
    "        best_params=[k,t,n_pca]\n",
    "        best_score=result\n",
    "    results+=[result]\n",
    "    print(f'parameters: {k,t,n_pca}')\n",
    "    print(f'Correlation After imputation: {result}\\n--------')\n",
    "\n",
    "\n",
    "print(f'Correlation before imputation: {baseline}')\n",
    "print(f'BEST CORRELATION AFTER IMPUTATION: {np.max(results)}')\n",
    "print(f'best parameters: {best_params}')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Differentiating by health status"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.9440258830507289"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# we apply the same procedure treating healthy and sick patients separately\n",
    "\n",
    "\n",
    "data_complete_healthy=pd.concat(patients_pp[1:3],axis=0)\n",
    "data_complete_sick=pd.concat([patients_pp[0],patients_pp[3]],axis=0)\n",
    "genes_healthy=list(data_complete_healthy.loc[:, (data_complete_healthy != data_complete_healthy.iloc[0]).any()].columns)\n",
    "genes_sick=list(data_complete_sick.loc[:, (data_complete_sick != data_complete_sick.iloc[0]).any()].columns)\n",
    "\n",
    "index_healthy=len(patients_pp[1])\n",
    "index_sick=len(patients_pp[0])\n",
    "\n",
    "\n",
    "imputed_global=data_complete_healthy.copy()\n",
    "imputed_local=sce.pp.magic(ad.AnnData(data_complete_healthy.loc[:,genes_healthy]),knn=80,t='auto',n_pca=175,n_jobs=8,copy=True,verbose=0,solver='approximate')\n",
    "imputed_global.loc[:,genes_healthy]=imputed_local.X\n",
    "result=0\n",
    "sc_imputed1= imputed_global.iloc[index_healthy:,:]\n",
    "sc_imputed2= imputed_global.iloc[:index_healthy,:]\n",
    "bulkified1=sc_imputed1.mean(axis=0)\n",
    "bulkified2=sc_imputed2.mean(axis=0)\n",
    "bulk1=bulk_train['patient' + str(2)]\n",
    "bulk2=bulk_train['patient' + str(3)]\n",
    "\n",
    "result=(spearmanr(bulkified1,bulk1).statistic + spearmanr(bulkified2,bulk2).statistic)/4\n",
    "result\n",
    "\n",
    "\n",
    "imputed_global=data_complete_sick.copy()\n",
    "imputed_local=sce.pp.magic(ad.AnnData(data_complete_sick.loc[:,genes_sick]),knn=80,t='auto',n_pca=175,n_jobs=8,copy=True,verbose=0,solver='approximate')\n",
    "imputed_global.loc[:,genes_sick]=imputed_local.X\n",
    "sc_imputed1= imputed_global.iloc[index_sick:,:]\n",
    "sc_imputed2= imputed_global.iloc[:index_sick,:]\n",
    "bulkified1=sc_imputed1.mean(axis=0)\n",
    "bulkified2=sc_imputed2.mean(axis=0)\n",
    "bulk1=bulk_train['patient' + str(1)]\n",
    "bulk2=bulk_train['patient' + str(4)]\n",
    "\n",
    "result+=(spearmanr(bulkified1,bulk1).statistic + spearmanr(bulkified2,bulk2).statistic)/4\n",
    "result\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Using also unlabelled data "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "parameters: (200, 'auto', 250)\n",
      "Correlation After imputation: 0.9524720550899409\n",
      "--------\n",
      "parameters: (200, 'auto', 300)\n",
      "Correlation After imputation: 0.9524691325148815\n",
      "--------\n",
      "parameters: (250, 'auto', 250)\n",
      "Correlation After imputation: 0.9525459803554612\n",
      "--------\n",
      "parameters: (250, 'auto', 300)\n",
      "Correlation After imputation: 0.9525375689279423\n",
      "--------\n",
      "Correlation before imputation: 0.9358073816280966\n",
      "BEST CORRELATION AFTER IMPUTATION: 0.9525459803554612\n",
      "best parameters: [250, 'auto', 250]\n"
     ]
    }
   ],
   "source": [
    "# we include also unlabelled data in the training set (i.e. patients 5,6,7)\n",
    "\n",
    "cells_5=[column for column in sc_test.columns if 'patient5' in column ]\n",
    "cells_6=[column for column in sc_test.columns if 'patient6' in column ]\n",
    "cells_7=[column for column in sc_test.columns if 'patient7' in column ]\n",
    "\n",
    "\n",
    "# we transpose the dataframe since we want a cells x genes dataframe for applying MAGIC\n",
    "sc5_train=sc_test[cells_5].T\n",
    "sc6_train=sc_test[cells_6].T\n",
    "sc7_train=sc_test[cells_7].T\n",
    "\n",
    "\n",
    "patients=[sc1_train,sc2_train,sc3_train,sc4_train,sc5_train,sc6_train,sc7_train]\n",
    "patients_pp=[]\n",
    "for i,data in enumerate(patients):\n",
    "    pp_data=pp(data)\n",
    "    patients_pp += [pp_data]\n",
    "data_complete=pd.concat(patients_pp,axis=0)\n",
    "genes=list(data_complete.loc[:, (data_complete != data_complete.iloc[0]).any()].columns)\n",
    "\n",
    "\n",
    "lens=[len(x) for x in patients_pp]\n",
    "indexes=[np.sum(lens[:i+1]) for i in range(4)]\n",
    "results=[]\n",
    "k_set ,t_set, n_pca_set= [200, 250], ['auto'] ,[250, 300]\n",
    "best_params=0\n",
    "best_score=0\n",
    "\n",
    "baseline=np.mean([spearmanr(patients_pp[i].mean(axis=0),bulk_train['patient' + str(i+1)]).statistic for i in range(4)])\n",
    "\n",
    "\n",
    "for k,t,n_pca in product(k_set,t_set,n_pca_set):\n",
    "    imputed_global=data_complete.copy()\n",
    "    imputed_local=sce.pp.magic(ad.AnnData(data_complete.loc[:,genes]),knn=k,t=t,n_pca=n_pca,n_jobs=8,copy=True,verbose=0,solver='approximate',random_state=7)\n",
    "    imputed_global.loc[:,genes]=imputed_local.X\n",
    "    result=0\n",
    "    for i,index in enumerate(indexes):\n",
    "        if i==0: sc_imputed= imputed_global.iloc[:index,:]\n",
    "        else: sc_imputed= imputed_global.iloc[indexes[i-1]:index,:]\n",
    "        bulkified=sc_imputed.mean(axis=0)\n",
    "        bulk=bulk_train['patient' + str(i+1)]\n",
    "        result+=spearmanr(bulkified,bulk).statistic/4\n",
    "    if result >= best_score:\n",
    "        best_params=[k,t,n_pca]\n",
    "        best_score=result\n",
    "    results+=[result]\n",
    "    print(f'parameters: {k,t,n_pca}')\n",
    "    print(f'Correlation After imputation: {result}\\n--------')\n",
    "\n",
    "\n",
    "print(f'Correlation before imputation: {baseline}')\n",
    "print(f'BEST CORRELATION AFTER IMPUTATION: {np.max(results)}')\n",
    "print(f'best parameters: {best_params}')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ML4G",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
